File: src/main/java/nz/co/searchwellington/views/GoogleMapsDisplayCleaner.java
package nz.co.searchwellington.views;

import java.util.List;

import nz.co.searchwellington.model.frontend.FrontendResource;

import org.apache.log4j.Logger;
import org.springframework.stereotype.Component;

import uk.co.eelpieconsulting.common.geo.DistanceMeasuringService;
import uk.co.eelpieconsulting.common.geo.model.Place;

import com.google.common.collect.Lists;

@Component
public class GoogleMapsDisplayCleaner {
        
    private static final double ONE_HUNDRED_METERS = 0.1;

	private static Logger log = Logger.getLogger(GoogleMapsDisplayCleaner.class);
    
    private final DistanceMeasuringService distanceMeasuringService;
    
    public GoogleMapsDisplayCleaner() {
		this.distanceMeasuringService = new DistanceMeasuringService();
	}
    
	public List<FrontendResource> dedupe(List<FrontendResource> geocoded) {
        return dedupe(geocoded, null);
    }

    public List<FrontendResource> dedupe(List<FrontendResource> geocoded, FrontendResource selected) {      
        log.debug("Deduping collection with " + geocoded.size() + " items");
        final List<FrontendResource> deduped = Lists.newArrayList();
        if (selected != null && selected.getPlace() != null) {
            deduped.add(selected);
        }
        
        for (FrontendResource resource : geocoded) {
        	if (resource.getPlace() != null) {
        		boolean isUnique = !listAlreadyContainsResourceWithThisLocation(deduped, resource);
        		if (isUnique) {
        			deduped.add(resource);
        		}
        	}
        }
        
        log.info("Returning collection with " + deduped.size() + " items");
        return deduped;
    }
    
    private boolean listAlreadyContainsResourceWithThisLocation(List<FrontendResource> deduped, FrontendResource candidiate) {
        for (FrontendResource resource : deduped) {
            if (areSameOrOverlappingLocations(resource.getPlace(), candidiate.getPlace())) {
                log.debug("Rejected " + candidiate.getName() + " as it overlaps more recent item: " + resource.getPlace().getAddress() + " / " + candidiate.getPlace().getAddress());
                return true;
            }
        }
        return false;
    }

	private boolean areSameOrOverlappingLocations(Place here, Place there) {
		final double distanceBetweenHereAndThere = distanceMeasuringService.getDistanceBetween(here.getLatLong(), there.getLatLong());
		return distanceBetweenHereAndThere < ONE_HUNDRED_METERS;
	}
    
}

File: src/main/java/nz/co/searchwellington/views/GoogleMapsDisplayCleaner.scala
package nz.co.searchwellington.views

import nz.co.searchwellington.model.{Geocode, Newsitem}
import org.apache.log4j.Logger
import org.springframework.stereotype.Component
import uk.co.eelpieconsulting.common.geo.DistanceMeasuringService
import uk.co.eelpieconsulting.common.geo.model.LatLong

@Component class GoogleMapsDisplayCleaner() {

  private val ONE_HUNDRED_METERS = 0.1
  private val log = Logger.getLogger(classOf[GoogleMapsDisplayCleaner])

  private val distanceMeasuringService = new DistanceMeasuringService

  def dedupe(geocoded: Seq[Newsitem], selected: Option[Newsitem] = None): Seq[Newsitem] = {
    log.debug("Deduping collection with " + geocoded.size + " items")
    var deduped = Seq(selected).flatten

    geocoded.map { resource =>
      resource.geocode.foreach { p =>
        val isUnique = !listAlreadyContainsResourceWithThisLocation(deduped, p)
        if (isUnique) deduped = deduped :+ resource
      }
    }
    log.debug("Returning collection with " + deduped.size + " items")
    deduped
  }

  private def listAlreadyContainsResourceWithThisLocation(deduped: Seq[Newsitem], geocode: Geocode): Boolean = {
    deduped.exists { r =>
      r.geocode.forall { rp =>
        areSameOrOverlappingLocations(rp, geocode)
      }
    }
  }

  private def areSameOrOverlappingLocations(here: Geocode, there: Geocode) = {
    def latLongFor(geocode: Geocode): Option[LatLong] = {
      for {
        lat <- geocode.latitude
        lon <- geocode.longitude
      } yield {
        new LatLong(lat, lon)
      }
    }

    for {
      hereLatLong <- latLongFor(here)
      thereLatLong <- latLongFor(there)
    } yield {
      val distanceBetweenHereAndThere = distanceMeasuringService.getDistanceBetween(hereLatLong, thereLatLong)
      distanceBetweenHereAndThere < ONE_HUNDRED_METERS
    }
  }.getOrElse(false)

}
